### ISSUE TITLE
type alias incorrectly flagged as unused when solely used in impl header

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#58593]
Consider the following code ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e36a771655d356700b8fcf9330292edd)):

```rust
type T = ();

struct S<X>(X); 

impl Clone for S<T> { fn clone(&self) -> Self { S(()) } }

fn main() {
    let s = S(());
    drop(s.clone());
}
```

Today this emits the following warning diagnostic:

```
warning: type alias is never used: `T`
 --> src/main.rs:1:1
  |
1 | type T = ();
  | ^^^^^^^^^^^^
  |
  = note: #[warn(dead_code)] on by default
```

But that type alias is not dead code. It is used in the `impl Clone for S<T> { ... }`, as one can see by trying to recompile the code after commenting out the type alias.

**Comment by oli-obk** [comment]
Previous instances of related (or possibly the same) problems:

* https://github.com/rust-lang/rust/issues/47131
* https://github.com/rust-lang/rust/issues/54234

**Comment by varkor** [comment]
I think this is essentially the same problem as https://github.com/rust-lang/rust/issues/59333#issuecomment-477820887.

**Comment by istankovic** [comment]
Seems to be fixed. Tested with 1.73 and got no warnings.

**Comment by zakrad** [comment]
This looks covereed by `tests/ui/lint/dead-code/alias-type-used-as-generic-arg-in-impl.rs` added in #141407 can be closed.
